Update README, add MLFlowDataSet, and create MinIO bucket automatically (#11)
deepyaman/kedro-streaming: README.md

> # kedro-streaming

Reviewer: ```suggestion
# Kedro-Streaming
```

## Overview

This repository serves as an introduction to Spark Streaming combined with Kedro. The infrastructure needed to enable

> streaming has all been defined in `docker-compose.yml`.

Reviewer: ```suggestion
streaming has all been defined in [`docker-compose.yml`](docker-compose.yml).
```

There are two pipelines defined, emulating a real-life scenario:

- `kedro run --pipeline train` - This pipeline takes care of data engineering, training a ML model, publishing it to

>   MLFlow, and running validation.

Reviewer: ```suggestion
  MLflow, and running validation.
```

- `kedro run --pipeline inference` - This pipeline is responsible for inference. It does data engineering on streaming

>   data, loads the ML model from MLFlow, and prints the outputted predictions as a stream to the console.

Reviewer: ```suggestion
  data, loads the ML model from MLflow, and prints the outputted predictions as a stream to the console.
```

>   (A Kafka Topic can also be configured)

Reviewer: ```suggestion
  A Kafka Topic can also be configured.
```

## Architecture

> ![Streaming Kedro](./streaming_kedro.png)

Reviewer: ```suggestion
![Kedro-Streaming](./streaming_kedro.png)
```

> We use Minio as an abstraction to a blob storage. Buckets that are defined here can be aliased to real S3 buckets for

Reviewer: ```suggestion
We use MinIO as an abstraction to a blob storage. Buckets that are defined here can be aliased to real S3 buckets for
```

> production scenarios. For the MLFlow metadata store, we use a `mysql` server. Kafka is used to handle the streaming

Reviewer: ```suggestion
production scenarios. For the MLflow metadata store, we use a MySQL server. Kafka is used to handle the streaming
```

architecture.

> ## Pre-requisites

Reviewer: ```suggestion
## Prerequisites
```

> * A Docker installation, and basic understanding of Docker and micro-services.

Reviewer: ```suggestion
* A Docker installation, and basic understanding of Docker and microservices
```

> * Kedro, and how Kedro Pipelines work

Reviewer: ```suggestion
* A high-level understanding of how Kedro pipelines work
```

> ## Step-by-step Instructions

Reviewer: ```suggestion
## Step-by-step instructions
```

> 1. Build the `mlflow_server` Docker Image.

Reviewer: ```suggestion
1. Build the `mlflow_server` Docker image.
```

```bash
docker build -t mlflow_server:latest -f Dockerfile.mlflow .
```

> 2. Start the Multi-container application.

Reviewer: ```suggestion
2. Start the multi-container application.
```

```bash
docker-compose up
```

**Services**

| Service    | Host      | Port  |
| ---------- | --------- | ----- |
| Kafka      | localhost | 19092 |

> | MLFlow     | localhost | 5000  |

Reviewer: ```suggestion
| MLflow     | localhost | 5000  |
```

| Minio (S3) | localhost | 9000  |

3. Create a new Conda Environment, and install all Python dependencies

```bash
conda create -n kedro-streaming python=3.7 && conda activate kedro-streaming && pip install -e src/
```

> 4. Run the Kedro Pipeline to train the model, and publish the model to the model registry in MLFlow.

Reviewer: ```suggestion
4. Run the Kedro pipeline to train the model, and publish the model to the MLflow Model Registry.
```

```bash
kedro run --pipeline train
```

> 5. Run the Kedro Pipeline for spinning up the inference engine, which will do predictions on data, as it comes in.

Reviewer: ```suggestion
5. Run the Kedro pipeline for spinning up the inference engine, which will do predictions on data, as it comes in.
```

>    The StreamingSparkDataset is defined to read from the `hello-fraud` Kafka topic. The pipeline is currently configured

Reviewer: ```suggestion
   The `SparkStreamingDataSet` is defined to read from the `hello-fraud` Kafka topic. The pipeline is currently configured
```
